### ISSUE TITLE
print output for metrics in verbose mode

### ISSUE THREAD

**Issue by kennethreitz** [heroku/buildpack-stdlib#27]

**Comment by kennethreitz** [comment]
add a sperate flag for this

**Comment by edmorley** [comment]
What environment variable name do you think would make sense for this?
`BUILDPACK_VERBOSE`
`BUILDPACK_METRICS`
...?

**Comment by jmorrell** [comment]
I don't think this is something that we would want to expose to users. What would be the benefit?

**Comment by edmorley** [comment]
My use cases are:
* "my deploys are slow, what is taking the most amount of time?"
* "how are our deploy times changing over time?"

There are currently no timestamps or durations at all in the log, other than the ones yarn outputs, and the "activity" page also doesn't list how long the entire build step took.

As-is I have to manually compare the timestamp in "activity" with when the IRC bot says "prod deployed" to work out how long it took (or use a stopwatch whilst tailing the build log; we use GitHub sync that doesn't trigger until CI complete, so not as easy as timing a `git push heroku master`).

The app in question uses both the nodejs and python buildpacks, so the time is eaten up somewhere in:
* nodejs binary installation
* yarn install (though yarn outputs its own duration, so I know this one is minimal)
* python binary installation (or cache restore)
* pip install (of whatever wasn't in the cache)
* python buildpack automatic django collectstatic step
* python buildpack `post_compile` custom script step (where we run a few things of our own)

From manual investigation I know that our slow steps are:
(a) django collectstatic (due to singlethreaded WhiteNoise compression; I have a WIP PR to make it multi-threaded),
(b) from the things we run in `post_compile`. 

However it would be good to be able to track those easily over time (by inspecting logs) to see whether our tweaks have helped.

I was thinking there would be two ways to quickly add this information to the log (given manual annotation is a pain / we can't easily annotate the parts of a buildpack outside `post_compile`):
1. Make the buildpack-stdlib's `puts_step` command prefix the output with a timestamp
2. Capitalise on the `mtime` metrics already recorded ([example](https://github.com/heroku/heroku-buildpack-python/blob/500daaf6fc58c7df6489582398d4f83021badf4b/bin/compile#L222))

**Comment by edmorley** [comment]
Related: heroku/heroku-buildpack-python#420 (since closed)

**Comment by jmorrell** [comment]
I'm definitely down with adding timing information to the logs. We could use `puts_verbose` for this, or create a `puts_timing` so that it's consistent. I don't think just blindly logging out the metrics is the best way to do this though.

[This is been an open issue on the Node buildpack since forever](https://github.com/heroku/heroku-buildpack-nodejs/issues/190). (my bad, goal for Q1 is to clear this backlog)

**Comment by edmorley** [comment]
> There are currently no timestamps or durations at all in the log, other than the ones yarn outputs, and the "activity" page also doesn't list how long the entire build step took.
> 
> As-is I have to manually compare the timestamp in "activity" with when the IRC bot says "prod deployed" to work out how long it took (or use a stopwatch whilst tailing the build log; we use GitHub sync that doesn't trigger until CI complete, so not as easy as timing a `git push heroku master`).

I've solved this via:
https://github.com/edmorley/heroku-buildpack-timestamps
